src/merger_tree.cpp

```diff
@@ -0,0 +1,175 @@
+    std::vector<HBTInt> all_trackids;
+    all_trackids.reserve(nr_particles);
+    std::vector<HBTInt> all_particle_ids;
+    all_particle_ids.reserve(nr_particles);
+    for(const auto &subhalo: Subhalos) {
+      if(subhalo.Nbound > 1) {
+        for(HBTInt i=0; i<subhalo.Nbound; i+=1) {
+          all_trackids.push_back(subhalo.TrackId);
+          all_particle_ids.push_back(subhalo.Particles[i].Id);
+        }
+      }
+    }
+    assert(all_trackids.size()==nr_particles);
+    assert(all_particle_ids.size()==nr_particles);
+  
+    // Count particles from subhalos which became unresolved
+    HBTInt nr_to_find = 0;
+    for (const auto &p : DescendantTracerIds) {
+      // Here p is (TrackId, vector of particle IDs) pair
+      nr_to_find += p.second.size();
+    }
+
+    // Populate vectors of particles to find
+    std::vector<HBTInt> find_particle_ids;
+    find_particle_ids.reserve(nr_to_find);
+    for (const auto &p : DescendantTracerIds) {
+      const HBTInt &LostTrackId = p.first;
+      const std::vector<HBTInt> &LostTracerIds = p.second;
+      for (const auto part_id : LostTracerIds) {
+        find_particle_ids.push_back(part_id);
+      }
+    }
+    assert(find_particle_ids.size()==nr_to_find);
+    
+    // Determine subhalo membership of the particle IDs to find:
+    //
+    // For each particle ID in find_particle_ids we identify all matching IDs in
+    // all_particle_ids (IDs of all bound subhalo particles) and return the TrackId
+    // of the subhalo that particle belongs to. Since HBT halos are not fully
+    // exclusive some particles may be found multiple times, possibly on other
+    // MPI ranks.
+    //
+    // count_found contains the number of times each entry in find_particle_ids was
+    // found.
+    //
+    // trackids_found contains the trackids of the matching particle IDs and has
+    // sum(count_found) entries.
+    LocateValuesById(all_particle_ids, all_trackids, MPI_HBT_INT, find_particle_ids,
+                     count_found, trackids_found, world.Communicator);
+  }
+
+  // Now identify a descendant for each lost subhalo in the DescendantTracerIds map
+  std::vector<std::pair<HBTInt,HBTInt>> DescendantTrackIds;
+  {
+    HBTInt count_found_offset = 0;
+    HBTInt trackids_found_offset = 0;
+    DescendantTrackIds.reserve(DescendantTracerIds.size());
+    for (const auto &p : DescendantTracerIds) {
+
+      // TrackId of the lost subhalo
+      const HBTInt &LostTrackId = p.first;
+    
+      // Vector of tracer particle IDs from the lost subhalo
+      const std::vector<HBTInt> &LostTracerIds = p.second;
+    
+      // Make a map where the keys are TrackIds of the possible descendants
+      // and the values are counts of how many particles went to each one.
+      std::map<HBTInt,HBTInt> nr_desc_particles;
+      // Loop over tracers to find for this lost subhalo
+      for(HBTInt i=0; i<LostTracerIds.size(); i+=1) {
+        // Loop over number of times each tracer was found
```

Comment:
Once we have removed the duplicate particles in the latest version, we will have a unique value per searched particle ID.